snakemake/executors.py

```diff
@@ -2093,3 +2107,641 @@ def run_wrapper(
+                        f, gb, warning_size_gb
+                    )
+                )
+            return f
+
+        self.workflow_sources = []
+
+        for wfs in self.workflow.get_sources():
+            if os.path.isdir(wfs):
+                for (dirpath, dirnames, filenames) in os.walk(wfs):
+                    self.workflow_sources.extend(
+                        [check_size(os.path.join(dirpath, f)) for f in filenames]
+                    )
+            else:
+                self.workflow_sources.append(check_size(os.path.abspath(wfs)))
+
+    def _generate_build_source_package(self):
+        """in order for the instance to access the working directory in storage,
+           we need to upload it. This file is cleaned up at the end of the run.
+           We do this, and then obtain from the instance and extract.
+        """
+        # We will generate a tar.gz package, renamed by hash
+        tmpname = next(tempfile._get_candidate_names())
+        targz = os.path.join(tempfile.gettempdir(), "snakemake-%s.tar.gz" % tmpname)
+        tar = tarfile.open(targz, "w:gz")
+
+        # Add all workflow_sources files
+        for filename in self.workflow_sources:
+            arcname = filename.replace(self.workflow.basedir + os.path.sep, "")
+            tar.add(filename, arcname=arcname)
+
+        tar.close()
+
+        # Rename based on hash, in case user wants to save cache
+        sha256 = get_file_hash(targz)
+        hash_tar = os.path.join(self.workdir, "snakeworkdir-%s.tar.gz" % sha256)
+
+        # Only copy if we don't have it yet, clean up if we do
+        if not os.path.exists(hash_tar):
+            shutil.move(targz, hash_tar)
+        else:
+            os.remove(targz)
+
+        # We will clean these all up at shutdown
+        self._build_packages.add(hash_tar)
+
+        return hash_tar
+
+    def _generate_job_action(self, job):
+        """generate a single action to execute the job.
+        """
+        # Derive the entrypoint command, the same content that might be written by self.get_jobscript(job)
+        use_threads = "--force-use-threads" if not job.is_group() else ""
+        exec_job = self.format_job(
+            self.exec_job, job, _quote_all=True, use_threads=use_threads
+        )
+
+        # The full command to download the archive, extract, and run
+        # For snakemake bases, we must activate the conda environment, but
+        # for custom images we must allow this to fail (hence || true)
+        commands = [
+            "/bin/bash",
+            "-c",
+            "mkdir -p /workdir && cd /workdir && wget -O /download.py https://gist.githubusercontent.com/vsoch/84886ef6469bedeeb9a79a4eb7aec0d1/raw/181499f8f17163dcb2f89822079938cbfbd258cc/download.py && chmod +x /download.py && source activate snakemake || true && python /download.py download %s %s /tmp/workdir.tar.gz && tar -xzvf /tmp/workdir.tar.gz && %s"
```

Comment:
The `|| true` looks suspicious. What is the reason?

Reply:
If they use a non traditional or custom container that doesn’t have an environment called snakemake it should still work.

Reply:
Actually, you shouldn't need the source activate snakemake at all. The official Snakemake docker container should have the envionment activated automatically.

Reply:
Indeed, I just tested this with the container and you are spot on! I'll test it with the pipelines first and then update the PR here upon success.

Reply:
hey @johanneskoester ! So I just tried this, and now I see why I added it (I think I must have started without it, because I couldn't have known that the container had the snakemake environment). If I don't do that source, then supporting libraries don't seem to be found. Here is the resulting error message:

```python
Traceback (most recent call last):
  File "/download.py", line 6, in <module>
    from google.cloud import storage
ModuleNotFoundError: No module named 'google.cloud'
Started running "snakejob-copy-2"
Stopped pulling "snakemake/snakemake:v5.7.2"
Started pulling "snakemake/snakemake:v5.7.2"
Worker "google-pipelines-worker-aa30ba7f3bbf98ad61c08b44d4d91fd4" assigned in "us-central1-a"
```
So I won't be updating the PR to remove that line, as it breaks function.

Reply:
Can you test this again, now that you use the latest container 5.10.0?